packages/harmony/src/components/input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,178 @@
+   */
+  const [isFocused, handleFocus, handleBlur] = useFocusState(
+    onFocusProp,
+    onBlurProp
+  )
+
+  // For focus behavior and accessiblity, <label> needs to have a htmlFor={} provided to an id matching the input
+  const backupId = useId()
+  const id = idProp ?? backupId
+
+  const characterCount = value !== undefined ? `${value}`.length : 0
+  const hasValue = characterCount > 0
+
+  // Hide the label when requested or when the size is set to small
+  const shouldShowLabel = !hideLabel && size !== TextInputSize.SMALL
+  const labelText = required ? `${labelProp} *` : labelProp
+  const helperTextSize: TextSize = size === TextInputSize.SMALL ? 'xs' : 's'
+
+  // Whenever a label isn't visible the placeholder should be visible in it's place (if provided)
+  const shouldShowPlaceholder = isFocused || !shouldShowLabel
+  const shouldShowAdornments = isFocused || hasValue || !shouldShowLabel
+  // Show the maxlength text whenever we hit a certain threshold (default 70%) + the input is focused
+  const shouldShowMaxLengthText =
+    isFocused &&
+    maxLength &&
+    characterCount >= showMaxLengthThreshold * maxLength
+  // Turn the maxlength text to the warning color whenever we hit a certain threshold (default 90%)
+  const showMaxlengthWarningColor =
+    maxLength && characterCount >= maxLengthWarningThreshold * maxLength
+
+  // Styles for the root of the input
+  const inputRootStyle = {
+    [styles.default]: size === TextInputSize.DEFAULT,
+    [styles.small]: size === TextInputSize.SMALL,
+    [styles.warning]: warningProp,
+    [styles.error]: error,
+    [styles.focused]: isFocused,
+    [styles.disabled]: disabled,
+    [styles.required]: required
+  }
+
+  // Styles for the input element itself
+  const inputElStyle = {
+    [styles.default]: size === TextInputSize.DEFAULT,
+    [styles.small]: size === TextInputSize.SMALL,
+    [styles.disabled]: disabled
+  }
+
+  const inputRender = (
+    <Flex
+      className={cn(styles.inputAdornmentRow, { [styles.focused]: isFocused })}
+    >
+      {startAdornmentText && shouldShowAdornments ? (
+        <Text variant='label' size='l' color='subdued'>
+          {startAdornmentText}
+        </Text>
+      ) : null}
+      <input
+        onFocus={handleFocus}
+        onBlur={handleBlur}
+        ref={ref}
+        className={cn(styles.inputElement, inputClassName, inputElStyle)}
+        value={value}
+        maxLength={maxLength}
+        disabled={disabled}
+        placeholder={shouldShowPlaceholder ? placeholder : undefined}
+        aria-label={labelText ?? props['aria-label']}
+        aria-required={required}
+        id={id}
+        {...other}
+      />
+      {endAdornmentText && shouldShowAdornments ? (
+        <Text variant='label' size='l' color='subdued'>
+          {endAdornmentText}
+        </Text>
+      ) : null}
+    </Flex>
+  )
+
+  return (
+    <Flex className={cn(styles.root, className)} direction='column'>
+      <label
+        htmlFor={id}
```

Comment:
We don't need htmlFor if the label wraps the input, right?

Reply:
While technically true, an explicit `for` is a best practice for accessibility since some screen readers can't handle implicit labels very well. 
Given that it's easy for us to do an explicit label here I don't see why we wouldn't 🤷 
![image](https://github.com/AudiusProject/audius-protocol/assets/6711655/73fd86e3-bca8-4218-b40b-1efc3eddf8fc)
https://www.w3.org/WAI/tutorials/forms/labels/#associating-labels-implicitly

Reply:
oof, fun. I didn't realize wrapping was problematic too 💀 
Not wrapping is trickier to implement though since we need the labels to wrap the "clickable" parts for focus management. Otherwise we would have to set up our own click handler to redirect focus.
Since a wrapped+explicit label "still works" according to them, I'm kind of inclined to just leave it as-is. Thoughts?

Reply:
SGTM